### PR TITLE
MAINT/DEP: properly implement `ndarray.__pos__`

### LINKED ISSUE #9081: ENH: call np.positive in ndarray.__pos__, and issue a deprecation warning when it falls
As I wrote in https://github.com/numpy/numpy/pull/8967#issuecomment-300273518:

> Probably, the desired behavior is something like:
```python
def __pos__(self):
    try:
        return np.positive(self)
    except TypeError:
        issue deprecation warning
        return self.copy()
```
> (This may be a little tricky, though, because it will need to be done in C.)

This will be useful for ndarray subclasses that override `__array_ufunc__`.

### PR DESCRIPTION
**Pull request by mhvk**
fixed #9081

For regular arrays, we now give a deprecation warning if np.positive cannot handle it; for ndarray subclasses that override __array_ufunc__, we pass on the type error.

### CODE REVIEW

File: numpy/core/src/multiarray/number.c [hunk 1]
```diff
@@ -550,6 +551,54 @@ array_power(PyArrayObject *a1, PyObject *o2, PyObject *modulo)
...
+     * 2. If we get an exception, check whether __array_ufunc__ is
+     *    overridden; if so, we live in the future and we allow the
+     *    TypeError to be passed on.
+     * 3. If not, give a deprecation warning and return a copy.
+     */
+    PyObject *value;
+    PyObject *mod;
+    if (mod) {
+        Py_DECREF(mod);
+    }
```
Comment (mattip) [inline numpy/core/src/multiarray/number.c]: This looks wrong. What is mod?

**Comment by mattip** [comment]
Why was this different, historically, than `array_negative` or `array_absolute`? It seems strange to special case only `array_positive`

File: numpy/core/src/multiarray/number.c (hunk 1 above)
Comment (mhvk) [reply numpy/core/src/multiarray/number.c]: Oops, leftover from debugging!! (I had the problem that `has_not_default...` doesn't work when an error is set)

**Comment by mhvk** [comment]
@mattip - I'm guessing it was just a short-cut. Certainly, `np.positive` didn't exist until recently.

File: numpy/core/src/private/ufunc_override.c [hunk 2]
```diff
@@ -61,7 +61,7 @@ get_non_default_array_ufunc(PyObject *obj)
  * Returns 1 if this is the case, 0 if not.
  */
 
-static int
+NPY_NO_EXPORT int
```
Comment (eric-wieser) [inline numpy/core/src/private/ufunc_override.c]: nit: This doesn't belong in this PR

Comment (eric-wieser) [reply numpy/core/src/private/ufunc_override.c]: edit: nevermind

**Comment by eric-wieser** [comment]
Commit tag should definitely contain `DEP`, since this has a deprecation

**Comment by mhvk** [comment]
OK, I changed the commit message to include "DEP".

File: numpy/core/src/multiarray/number.c [hunk 3]
```diff
@@ -550,6 +551,50 @@ array_power(PyArrayObject *a1, PyObject *o2, PyObject *modulo)
...
+         */
+        PyObject *exc, *val, *tb;
+        PyErr_Fetch(&exc, &val, &tb);
+        if (has_non_default_array_ufunc((PyObject *)m1)) {
+            PyErr_Restore(exc, val, tb);
+            return NULL;
+        }
+        /* 2018-06-28, 1.16.0 */
+        if (DEPRECATE("Applying '+' to a non-numerical array is "
+                      "ambigous. Returning a copy, but in the future "
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: Typo: `ambiguous`.

Is this even true? It's just ill-defined, not ambiguous.

**Review by mattip — APPROVED** [review]

**Review by eric-wieser — CHANGES_REQUESTED** [review]
Just drawing attention to my comment above

File: numpy/core/src/multiarray/number.c (hunk 3 above)
Comment (mhvk) [reply numpy/core/src/multiarray/number.c]: OK, replaced with "ill-defined".

**Comment by mattip** [comment]
@eric-wieser could you comment on the changes you requested?

@mhvk it seems a rebase is needed

**Review by eric-wieser — APPROVED** [review]

**Comment by mhvk** [comment]
@mattip - thanks for noticing... now rebased and approved, so can go in if tests pass.
